File: 3rd Year/5th Semester/Design and Analysis of Algorith Lab/Day-01/Q2.java
// Write a program to find out the second smallest and second largest element stored in an array of n integers. 
// Input: Size of the array is ‘n’ and read ‘n’ number of elements from a disc file.

import java.util.*;
public class Q2 {


    private static int secondLargest(int[] arr) {
        int largest = arr[0];
        int secondLargest = -1;

        for(int i = 0; i < arr.length; i++) {
            if(arr[i] > largest) {
                secondLargest = largest;
                largest = arr[i];
            }else if(arr[i] < largest && arr[i] > secondLargest) {
                secondLargest = arr[i];
            }
        }

        return secondLargest;
    }

    private static int secondSmallest(int[] arr) {
        int smallest = Integer.MAX_VALUE;
        int secondSmallest = arr[0];

        for(int i = 0; i < arr.length; i++) {
            if(arr[i] < smallest) {
                secondSmallest = smallest;
                smallest = arr[i];
            }else if(arr[i] > smallest && arr[i] < secondSmallest) {
                secondSmallest = arr[i];
            }
        }

        return secondSmallest;
    }
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter size of array: ");
        int n = sc.nextInt();
        System.out.println("Enter elements");
        int []arr = new int[n];
        for(int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }


        System.out.println("Second Largest: " + secondLargest(arr) + " Second Smallest: " + secondSmallest(arr));

        sc.close();
    }
}

File: 3rd Year/5th Semester/Design and Analysis of Algorith Lab/Day-02/q2.c
// Write a program in C to find GCD of two numbers using recursion.

#include<stdio.h>
int gcd(int a, int b) {
    if(b == 0) {
        return a;
    }
    return gcd(b, a % b);
}

int main() {
    int n;
    printf("Enter the number of pairs: ");
    scanf("%d", &n);
    int arr[n * 2];
    printf("Enter the pairs: ");
    for(int i = 0; i < n * 2; i++) {
        scanf("%d", &arr[i]);
    }
    
    printf("All the pairs\n");
    for(int i = 0; i < n * 2; i += 2) {
        printf("GCD of %d and %d is: %d\n", arr[i], arr[i + 1], gcd(arr[i], arr[i + 1]));
    }
    return 0;
}